design-proposals/VM-cloning.md

```diff
@@ -0,0 +1,119 @@
+
+## Motivation
+The ability to clone a VM is a generic virtualization operation which may be useful for several use-cases.
+
+For example, a system admin / user might want to base a new VM on an existing one. In such a case we can clone a VM and
+then change it according to our needs. In addition, a cluster admin might want to try different configurations
+(for example, to check whether other configuration provide better performance for his workload), in this case he can
+clone the existing VM, test it and decide whether to move the new settings.
+
+Naively copy/pasting a VM spec would not work in most cases therefore this serves as an automation for cloning safely.
+
+## Goals
+Have a mechanism to clone VMs. Original and cloned VM would have the same:
+* Disks
+* Network configuration
+* Volumes
+* CPU topology
+
+Also, provide an ability to fine-tune what's being cloned. For example, it is possible to clone or skip
+annotations / labels.  
+
+## Non Goals
+In first implementation iteration the following limitations exist:
+* Only offline cloning will be supported, meaning cloning a VM when it's stopped.
+* Cloned VM would be in the same namespace as the original one.
+* Host disks would not be supported.
+
+## Definition of Users
+Everyone who creates a VM might create a VM clone.
+
+## User Stories
+* As a user / admin, I want to be able to create a new VM that is based on an existing VM.
+* As a user / admin, I want to create a clone to a VM using virtctl.
+
+## Repos
+Kubevirt/kubevirt
+
+# Design
+Design can be inspired by the way VMIs are being migrated. In order to migrate a VMI, a [VirtualMachineInstanceMigration](https://kubevirt.io/user-guide/operations/live_migration/#initiate-live-migration)
+CRD is posted into the cluster. This CRD is very thin and basically triggers a migration with basic parameters.
+
+For VM cloning a similar approach can be taken: a new CRD would be introduced that would provide some basic parameters like:
+* The VM that needs to be cloned
+* A new name for the VM (could be optional - a new name will be generated)
+* New MAC address (optional: if not provided MAC would be erased)
+* Whether to clone or skip labels / annotations etc.
+
+More parameters could be added in the future, like the new namespace for the VM, and more.
+
+## API Examples
+As said above, API can be similar to [VirtualMachineInstanceMigration](https://kubevirt.io/user-guide/operations/live_migration/#initiate-live-migration).
+
+Here's a sketch for the new CRD:
+```yaml
+kind: VirtualMachineClone
+  spec:
+    source:
+      # More sources may be supported in the future, e.g. VirtualMachinInstance or VirtualMachineSnapshot
+      kind: VirtualMachine
+      name: my-other-cool-vm
+    
+    targetVirtualMachineName: my-cool-clone   # Optional - a new name can be generated, e.g. my-cool-vm-jfh54b 
+    newMacAddress: my-new-mac-address         # Optional - new mac address can be generated automatically
```

Comment:
I see only one field here, but there may be multiple interfaces to a VM.

Reply:
So every interface has its own MAC address?
If so - WDYT on having a way to specify a new MAC per interface (and reference it by name)?

Reply:
Cool idea!